core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/PlainHighlighter.java

```diff
@@ -104,9 +105,14 @@ public HighlightField highlight(HighlighterContext highlighterContext) {
 
         try {
             textsToHighlight = HighlightUtils.loadFieldValues(field, mapper, context, hitContext);
-
             for (Object textToHighlight : textsToHighlight) {
-                String text = textToHighlight.toString();
+                String text;
+                if (textToHighlight instanceof BytesRef) {
+                    // keywords are internally stored as utf8 bytes
+                    text = ((BytesRef) textToHighlight).utf8ToString();
+                } else {
+                    text = textToHighlight.toString();
+                }
```

Comment:
I agree that it's better to call valueForDisplay (I'll change the PR) though we are protected here since the highlighting can be done only on text or keyword fields.